### ISSUE TITLE
Pip args in Dockerfile not exposed via containerize command

### ISSUE THREAD

**Issue by gregd33** [bentoml/BentoML#1036]
**Describe the bug**

I believe this is a bug, but I might just be missing how to properly use the containerize functionality. 

The Dockerfile generated by https://github.com/bentoml/BentoML/blob/master/bentoml/clipper/__init__.py#L97 includes two args for PIP repos. However, the containerize  functionality doesn't seem to expose these arguments. The problem is that this will override other pip config in the image (e.g. if I build on a base image that has pip configured with repos I want, these environment variables will override it). 

So the containerize function essentially forces the default pip repo. 

**To Reproduce**
1. Use a base image with a custom pip config other than the default
2. Build a service with this image and containerize

**Expected behavior**
Your pip config will be be ignored in favour of the default args in the Dockerfile
(logging.DEBUG)

**Additional context**
This is mostly relevant when building in environments without access to the main pypi repo

**Potential fix** 

Either add args specifically for the PIP build args or allow for an arbitrary list of build args.

**Comment by yubozhao** [comment]
@gregd33 Thanks for reporting this. We missed providing options for those args in the `bentoml containerize` command. 

You can utilize the docker command as a workaround for the moment.  If your BentoService is saved in a remote Yatai service, you can first use `bentoml retrieve my_service:my_version` to download the BentoService to the current directory and then run docker build command.

If you have any bandwidth, we would love to have your contribution.

**Comment by parano** [comment]
Hey @gregd33, thanks for reporting this. Indeed the following line in the generated docker image will override your PyPI config in the base image, and it requires the user to specify the build args again, which is not necessary if it is already specified in the docker base image.

```
ARG PIP_INDEX_URL=https://pypi.python.org/simple/
ARG PIP_TRUSTED_HOST=pypi.python.org
ENV PIP_INDEX_URL $PIP_INDEX_URL
ENV PIP_TRUSTED_HOST $PIP_TRUSTED_HOST
```

On a side note, the dockerfile is actually generated from: https://github.com/bentoml/BentoML/blob/v0.8.6/bentoml/saved_bundle/templates.py#L79 and it is similar to the clipper base image you've pointed to.

We are thinking about making the `pip_index_url` and `pip_trusted_host` as two additional arguments passed to the `@env` decorator, e.g.:

```python
@env(
  auto_pip_dependencies=True,
  pip_index_url='my_pypi_host_url',
  pip_trusted_host='my_pypi_host_url'
)
@artifacts([SklearnModelArtifact('model')])
class IrisClassifier(BentoService):
  ...
```
This will allow us to move the logic of applying PyPI config from bash script `bentoml-init.sh`, and avoid the overriding issue.

And yes the suggestion to add arbitrarily `--build-args` to the `bentoml containerize` command and pass it down to the `docker build` API call internally also makes a lot of sense and can be very useful for other build-args provided by the custom docker base image.

As a workaround for now, you can always build with the `docker build` command with the saved bundle directory and pass the `--build-arg`  to avoid overriding with the default:
```bash
docker build --build-arg PIP_INDEX_URL=http://127.0.0.1:8000/pypi --build-arg PIP_TRUSTED_HOST=127.0.0.1:8000 -t myimage:v1  $saved_path 
```

**Comment by gregd33** [comment]
@parano - thanks, putting them in `@env` makes a lot of sense and would work perfectly. 

And thanks for the workaround - that's what I had been doing so far.  But in a long run, being able to use the `containerize` command would be ideal.

**Comment by danield137** [comment]
@parano , can you also consider adding `pip_extra_index_url` as a list? this will allow users to use a private PyPI for a subset of the packages

**Comment by parano** [comment]
@danield137 didn't know that option exists, thanks for the suggestion! we should definitely add that!

**Comment by danield137** [comment]
@parano didn't know too until I had to use it 🤣

**Comment by parano** [comment]
Both the original issue and the proposal on adding extra pip index parameters are added in PR https://github.com/bentoml/BentoML/pull/1084

It will be available in the next release 0.9.0
